File: src/pager.rs
use bytes::BytesMut;
use std::fs::{File, OpenOptions};
use std::io::{
    self,
    prelude::{Read, Seek, Write},
    SeekFrom,
};
use std::path::Path;

use crate::ChiError;

pub const PAGE_SIZE: usize = 4096 * 4; // 8 Kb
pub const HEADER_SIZE: usize = 100;

/// Represents a in-memory copy of page
pub struct MemPage {
    /// Number of physical page
    pub n_page: u32,

    /// Offset where to start to read from data
    pub offset: u16,

    /// Page bytes data
    data: BytesMut,
}

impl MemPage {
    /// Create a new MemPage
    pub fn new(n_page: u32, raw: [u8; PAGE_SIZE], offset: u16) -> Self {
        let mut data = BytesMut::with_capacity(raw.len());
        data.extend_from_slice(&raw);
        MemPage {
            n_page,
            data,
            offset,
        }
    }

    /// Return a reference to the data of page starting from offset
    pub fn data(&self) -> &[u8] {
        &self.data[self.offset as usize..]
    }

    /// Like `data` but return a mutable reference
    pub fn data_as_mut(&mut self) -> &mut [u8] {
        &mut self.data[self.offset as usize..]
    }

    /// Return a reference to all data from page
    pub fn raw(&self) -> &[u8] {
        &self.data[..]
    }

    /// Update raw data of page
    ///
    /// This function override current in-memory data of page
    /// to the new raw data from param taking care to update data
    /// using the offset of page as base.
    pub fn set_data(&mut self, raw: &[u8]) {
        let mut data = BytesMut::with_capacity(self.data.capacity());
        data.extend_from_slice(&self.data[..self.offset as usize]);
        data.extend_from_slice(raw);
        self.data = data;
    }
}

#[derive(Debug)]
pub struct Pager {
    buffer: File,
    total_pages: u32,
}

impl Pager {
    /// Open a file
    //
    // This function opens a file for paged access.
    //
    // Parameters
    // - filename: Database file (might not exist)
    pub fn open(filename: &Path) -> Result<Pager, ChiError> {
        let buffer = OpenOptions::new()
            .read(true)
            .write(true)
            .create(true)
            .open(filename)?;
        Ok(Pager {
            buffer,
            total_pages: 0,
        })
    }

    /// Read the file header
    ///
    /// This function reads in the header of a chidb file and returns it
    /// in a byte array. Note that this function can be called even if
    /// the page size is unknown, since the chidb header always occupies
    /// the first 100 bytes of the file.
    pub fn read_header(&mut self) -> Result<BytesMut, ChiError> {
        self.buffer.seek(SeekFrom::Start(0))?;
        let mut header = [0; HEADER_SIZE];
        self.buffer.read(&mut header)?;
        let mut bytes = BytesMut::with_capacity(header.len());
        bytes.extend_from_slice(&header);
        Ok(bytes)
    }

    pub fn write_header(&mut self, header: &[u8; HEADER_SIZE]) -> Result<(), ChiError> {
        self.buffer.seek(SeekFrom::Start(0))?;
        self.buffer.write(header)?;
        Ok(())
    }

    /// Allocate an extra page on the file and returns the page number
    pub fn allocate_page(&mut self) -> u32 {
        // We simply increment the page number counter.
        // read_page and write_page take care of the rest.
        //
        // TODO: We should create an empty page here???
        self.total_pages += 1;
        self.total_pages
    }

    /// Read a page from file
    ///
    /// This page reads a page from the file, and creates an in-memory copy
    /// in a MemPage struct (see header file for more details on this struct).
    /// Any changes done to a MemPage will not be effective until you call
    /// chidb_Pager_writePage with that MemPage.
    ///
    /// Parameters
    /// - n_page: Page number of page to read.
    pub fn read_page(&mut self, n_page: u32) -> Result<MemPage, ChiError> {
        if n_page > self.total_pages || n_page <= 0 {
            return Err(ChiError::EPageNo);
        }
        let seek = (n_page - 1) * PAGE_SIZE as u32;
        self.buffer.seek(SeekFrom::Start(seek as u64))?;

        let mut data = [0; PAGE_SIZE];
        let count = self.buffer.read(&mut data)?;
        println!("Read {} bytes from page {}", count, n_page);

        // Page one is special, the first HEADER_SIZE are used by the header
        // so we start to read after the header.
        // http://chi.cs.uchicago.edu/chidb/fileformat.html#physical-organization
        let mut offset = 0;
        if n_page == 1 {
            offset = HEADER_SIZE as u16 + 1;
        }

        Ok(MemPage::new(n_page, data, offset))
    }

    pub fn write_page(&mut self, page: &MemPage) -> Result<(), ChiError> {
        if page.n_page > self.total_pages || page.n_page <= 0 {
            return Err(ChiError::EPageNo);
        }
        let seek = (page.n_page - 1) * PAGE_SIZE as u32;
        self.buffer.seek(SeekFrom::Start(seek as u64))?;
        let count = self.buffer.write(&page.data)?;
        println!("Wrote {} bytes to page {}", count, page.n_page);
        Ok(())
    }

    pub fn is_empty(&self) -> Result<bool, io::Error> {
        let size = self.buffer.metadata()?.len();
        Ok(size == 0)
    }
}

File: go/pager.go
package chidb

import (
	"bytes"
	"errors"
	"fmt"
	"io"
	"log"
	"os"
)

const (
	PageSize   = 4096 * 4 // 8 Kb
	HeaderSize = 100
)

var ErrIncorrectPageNumber = errors.New("incorrect page number")

// MemPage Represents a in-memory copy of page
type MemPage struct {

	// Number of physical number
	number uint32

	// Offset where to start to read or write on data
	offset uint16

	// Page bytes data
	data [PageSize]byte
}

// Read returns the bytes of the page
// The returned data is only data avaliable to write and read in page
func (m *MemPage) Read() []byte {
	return m.data[m.offset:]
}

// WriteAt write data on page after at value
func (m *MemPage) WriteAt(data []byte, at uint16) error {
	buffer := bytes.NewBuffer([]byte(""))
	buffer.Grow(PageSize)

	dataSize := uint16(len(m.data))

	if l := dataSize; l < at {
		return fmt.Errorf("page data %d is less than %d", l, at)
	}

	// Write data that is before of `at` value
	if _, err := buffer.Write(m.data[:at]); err != nil {
		return err
	}

	// Write new data
	writen, err := buffer.Write(data)
	if err != nil {
		return err
	}

	remaning := at + uint16(writen)

	if remaning < dataSize {
		// Write the remaning bytes
		if _, err := buffer.Write(m.data[remaning:]); err != nil {
			return err
		}
	}

	if uint16(buffer.Len()) != dataSize {
		fmt.Printf("Buffer len: %d\n", buffer.Len())
		fmt.Printf("Page len: %d\n", dataSize)
		fmt.Println("-----------------------")
		panic("something goes really wrong here")
	}

	newData := buffer.Bytes()
	copy(m.data[:], newData[:PageSize])

	return nil
}

// Write write data on current page
// NOTE: the data param should has the same size of PageSize
func (m *MemPage) Write(data []byte) error {
	buffer := bytes.NewBuffer([]byte(""))
	buffer.Grow(PageSize)

	if _, err := buffer.Write(m.data[:m.offset]); err != nil {
		return err
	}

	if _, err := buffer.Write(data); err != nil {
		return err
	}

	if l := buffer.Len(); l != PageSize {
		return fmt.Errorf("invalid page size to write: expected %d got %d", PageSize, l)
	}

	newData := buffer.Bytes()
	copy(m.data[:], newData[:PageSize])

	return nil
}

// Len returns the lenght of page data available to read and write
func (m *MemPage) Len() int {
	return len(m.Read())
}

type Pager struct {
	buffer     *os.File
	totalPages uint32
}

// OpenPager opens a file for paged access
func OpenPager(filename string) (*Pager, error) {
	f, err := os.OpenFile(filename, os.O_CREATE|os.O_RDWR, os.ModePerm)
	if err != nil {
		return nil, err
	}

	return &Pager{
		buffer:     f,
		totalPages: 0,
	}, nil
}

// ReadHeader reads in the header of a chidb file and returns it
// in a byte array. Note that this function can be called even if
// the page size is unknown, since the chidb header always occupies
// the first 100 bytes of the file.
func (p *Pager) ReadHeader() ([]byte, error) {
	if _, err := p.buffer.Seek(0, io.SeekStart); err != nil {
		return nil, err
	}

	header := make([]byte, HeaderSize)
	if _, err := p.buffer.Read(header); err != nil {
		return nil, err
	}

	return header, nil
}

func (p *Pager) WriteHeader(header []byte) error {
	if _, err := p.buffer.Seek(0, io.SeekStart); err != nil {
		return err
	}

	if l := len(header); l != HeaderSize {
		return fmt.Errorf("invalid header size %d", l)
	}

	if _, err := p.buffer.Write(header); err != nil {
		return err
	}
	return nil
}

// ReadPage read a page from file
// This page reads a page from the file, and creates an in-memory copy
// in a MemPage struct (see header file for more details on this struct).
// Any changes done to a MemPage will not be effective until you call
// chidb_Pager_writePage with that MemPage.
func (p *Pager) ReadPage(page uint32) (*MemPage, error) {
	if err := p.pageIsValid(page); err != nil {
		return nil, err
	}

	var data [PageSize]byte
	count, err := p.buffer.ReadAt(data[:], p.offset(page))
	if err != nil {
		if !errors.Is(err, io.EOF) {
			return nil, fmt.Errorf("read buffer: %w", err)
		}
	}
	log.Printf("Read %d bytes from page %d\n", count, page)

	// Page one is special, the first `HeaderSize` are used by the header
	// so we start to read after the header.
	// http://chi.cs.uchicago.edu/chidb/fileformat.html#physical-organization
	offset := uint16(0)
	if page == 1 {
		offset = HeaderSize
	}

	return &MemPage{
		number: page,
		data:   data,
		offset: offset,
	}, nil
}

// WritePage write a page to file
// This page writes the in-memory copy of a page (stored in a MemPage
// struct) back to disk.
func (p *Pager) WritePage(page *MemPage) error {
	if err := p.pageIsValid(page.number); err != nil {
		return err
	}

	if l := len(page.data); l != PageSize {
		return fmt.Errorf("invalid page data size: expected %d got %d", PageSize, l)
	}

	offset := p.offset(page.number)
	count, err := p.buffer.WriteAt(page.data[:], offset)
	if err != nil {
		return err
	}
	log.Printf("Wrote %d bytes to page %d\n", count, page.number)

	return nil
}

// AllocatePage Allocate an extra page on the file and returns the page number
func (p *Pager) AllocatePage() uint32 {
	// We simply increment the page number counter.
	// ReadPage and WritePage take care of the rest.
	p.totalPages += 1
	return p.totalPages
}

func (p *Pager) IsEmpty() (bool, error) {
	info, err := p.buffer.Stat()
	if err != nil {
		return false, err
	}
	return info.Size() == 0, nil
}

func (p *Pager) Close() error {
	return p.buffer.Close()
}

func (p *Pager) pageIsValid(page uint32) error {
	if page > p.totalPages || page <= 0 {
		return ErrIncorrectPageNumber
	}
	return nil
}

func (p *Pager) offset(page uint32) int64 {
	return int64((page - 1) * PageSize)
}
